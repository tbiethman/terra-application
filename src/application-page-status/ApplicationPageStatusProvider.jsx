import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Scroll from 'terra-scroll';
import StatusView from 'terra-status-view';
import ApplicationPageStatusContext from './ApplicationPageStatusContext';
import styles from './ApplicationPageStatusProvider.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The components to be rendered within the context of the ApplicationPageStatusProvider.
   * Components rendered here are able to interact with ApplicationPageStatusProvider through
   * the ApplicationPageStatusContext.
   */
  children: PropTypes.node,
  /**
   * A function to be called with the current ref of the scrollable element rendered within the
   * ApplicationPageStatusProvider.
   */
  scrollRefCallback: PropTypes.func,
};

const ApplicationPageStatusProvider = ({ children, scrollRefCallback, ...customProps }) => {
  const [registeredPageStatus, setRegisteredPageStatus] = React.useState({});

  const contextValue = useMemo(() => ({
    show: (key, data) => {
      setRegisteredPageStatus(state => (
        {
          ...state,
          [`${key}`]: data,
        }
      ));
    },
    hide: (key) => {
      setRegisteredPageStatus((state) => {
        const newRegisteredPageStatus = { ...state };
        delete newRegisteredPageStatus[key];

        return newRegisteredPageStatus;
      });
    },
  }), []);

  const registeredPageStatusKeys = Object.keys(registeredPageStatus);
  const registeredButtonAttrs = registeredPageStatusKeys.map(key => (registeredPageStatus[key] && registeredPageStatus[key].buttonAttrs));
  const registeredMessage = registeredPageStatusKeys.map(key => (registeredPageStatus[key] && registeredPageStatus[key].message));
  const registeredTitle = registeredPageStatusKeys.map(key => (registeredPageStatus[key] && registeredPageStatus[key].title));
  const registeredVariant = registeredPageStatusKeys.map(key => (registeredPageStatus[key] && registeredPageStatus[key].variant));

  let className = cx('container');
  if (customProps.className) {
    className = [className, customProps.className].join(' ');
  }

  // In the event when multiple ApplicationPageStatus's are provided, the last rendered wins
  const lastIndex = registeredPageStatusKeys.length - 1;
  const statusView = (lastIndex >= 0) && (
    <StatusView
      buttonAttrs={registeredButtonAttrs[lastIndex]}
      className={cx('status-view')}
      message={registeredMessage[lastIndex]}
      title={registeredTitle[lastIndex]}
      variant={registeredVariant[lastIndex]}
    />
  );

  return (
    <div
      {...customProps}
      className={className}
    >
      {statusView}
      <Scroll refCallback={scrollRefCallback}>
        <ApplicationPageStatusContext.Provider value={contextValue}>
          {children}
        </ApplicationPageStatusContext.Provider>
      </Scroll>
    </div>
  );
};

ApplicationPageStatusProvider.propTypes = propTypes;

export default ApplicationPageStatusProvider;
