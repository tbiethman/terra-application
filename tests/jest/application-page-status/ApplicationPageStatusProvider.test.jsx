import React from 'react';
import ApplicationPageStatusProvider from '../../../src/application-page-status/ApplicationPageStatusProvider';

describe('ApplicationPageStatusProvider', () => {
  describe('Snapshots', () => {
    it('should render with minimal props', () => {
      const wrapper = shallow((
        <ApplicationPageStatusProvider />
      ));

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with custom props', () => {
      const wrapper = shallow((
        <ApplicationPageStatusProvider className="testClassName" />
      ));

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with optional props', () => {
      const scrollRefCallback = jest.fn();
      const wrapper = shallow((
        <ApplicationPageStatusProvider
          scrollRefCallback={scrollRefCallback}
        >
          <div>Test child</div>
        </ApplicationPageStatusProvider>
      ));

      expect(wrapper).toMatchSnapshot();
    });

    it('should render the status view based on the current state', () => {
      const reactUseState = React.useState;
      const mockSetState = jest.fn();

      const StatusViewButtons = [
        {
          text: 'Action 1',
          key: 1,
          size: 'medium',
        }, {
          text: 'Action 2',
          key: 2,
          size: 'medium',
        },
      ];

      React.useState = () => ([{
        1: {
          buttonAttrs: StatusViewButtons,
          message: 'Status View with all props specified',
          title: 'Jest Test',
          variant: 'no-data',
        },
      }, mockSetState]);

      const wrapper = shallow((
        <ApplicationPageStatusProvider>
          <div>Test child</div>
        </ApplicationPageStatusProvider>
      ));

      expect(wrapper).toMatchSnapshot();

      React.useState = reactUseState;
    });

    it('should render the status view when only buttonAttrs is specified', () => {
      const reactUseState = React.useState;
      const mockSetState = jest.fn();

      const StatusViewButtons = [
        {
          text: 'Action 1',
          key: 1,
          size: 'medium',
        },
      ];

      React.useState = () => ([{
        1: { buttonAttrs: StatusViewButtons },
      }, mockSetState]);

      const wrapper = shallow((
        <ApplicationPageStatusProvider>
          <div>Test child</div>
        </ApplicationPageStatusProvider>
      ));

      expect(wrapper).toMatchSnapshot();

      React.useState = reactUseState;
    });

    it('should render the status view when only message is specified', () => {
      const reactUseState = React.useState;
      const mockSetState = jest.fn();

      React.useState = () => ([{
        1: { message: 'Status View with only message specified' },
      }, mockSetState]);

      const wrapper = shallow((
        <ApplicationPageStatusProvider>
          <div>Test child</div>
        </ApplicationPageStatusProvider>
      ));

      expect(wrapper).toMatchSnapshot();

      React.useState = reactUseState;
    });

    it('should render the status view when only title is specified', () => {
      const reactUseState = React.useState;
      const mockSetState = jest.fn();

      React.useState = () => ([{
        1: { title: 'Jest Test' },
      }, mockSetState]);

      const wrapper = shallow((
        <ApplicationPageStatusProvider>
          <div>Test child</div>
        </ApplicationPageStatusProvider>
      ));

      expect(wrapper).toMatchSnapshot();

      React.useState = reactUseState;
    });

    it('should render the status view when only variant is specified', () => {
      const reactUseState = React.useState;
      const mockSetState = jest.fn();

      React.useState = () => ([{
        1: { variant: 'error' },
      }, mockSetState]);

      const wrapper = shallow((
        <ApplicationPageStatusProvider>
          <div>Test child</div>
        </ApplicationPageStatusProvider>
      ));

      expect(wrapper).toMatchSnapshot();

      React.useState = reactUseState;
    });

    it('should render a single status view with the configuration from the last rendered ApplicationPageStatus when more than one are specified', () => {
      const reactUseState = React.useState;
      const mockSetState = jest.fn();

      React.useState = () => ([{
        1: { message: 'First message', variant: 'error' },
        2: { message: 'Second message', variant: 'not-authorized' },
        3: { message: 'Third message', variant: 'no-data' },
      }, mockSetState]);

      const wrapper = shallow((
        <ApplicationPageStatusProvider>
          <div>Test child</div>
        </ApplicationPageStatusProvider>
      ));

      expect(wrapper).toMatchSnapshot();

      React.useState = reactUseState;
    });
  });
});
