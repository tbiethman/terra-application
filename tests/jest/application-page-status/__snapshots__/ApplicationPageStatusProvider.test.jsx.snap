// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationPageStatusProvider Snapshots should render a single status view with the configuration from the last rendered ApplicationPageStatus when more than one are specified 1`] = `
<div
  className="container"
>
  <StatusView
    buttonAttrs={Array []}
    className="status-view"
    isAlignedTop={false}
    isGlyphHidden={false}
    message="Third message"
    variant="no-data"
  />
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render the status view based on the current state 1`] = `
<div
  className="container"
>
  <StatusView
    buttonAttrs={
      Array [
        Object {
          "key": 1,
          "size": "medium",
          "text": "Action 1",
        },
        Object {
          "key": 2,
          "size": "medium",
          "text": "Action 2",
        },
      ]
    }
    className="status-view"
    isAlignedTop={false}
    isGlyphHidden={false}
    message="Status View with all props specified"
    title="Jest Test"
    variant="no-data"
  />
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render the status view when only buttonAttrs is specified 1`] = `
<div
  className="container"
>
  <StatusView
    buttonAttrs={
      Array [
        Object {
          "key": 1,
          "size": "medium",
          "text": "Action 1",
        },
      ]
    }
    className="status-view"
    isAlignedTop={false}
    isGlyphHidden={false}
  />
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render the status view when only message is specified 1`] = `
<div
  className="container"
>
  <StatusView
    buttonAttrs={Array []}
    className="status-view"
    isAlignedTop={false}
    isGlyphHidden={false}
    message="Status View with only message specified"
  />
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render the status view when only title is specified 1`] = `
<div
  className="container"
>
  <StatusView
    buttonAttrs={Array []}
    className="status-view"
    isAlignedTop={false}
    isGlyphHidden={false}
    title="Jest Test"
  />
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render the status view when only variant is specified 1`] = `
<div
  className="container"
>
  <StatusView
    buttonAttrs={Array []}
    className="status-view"
    isAlignedTop={false}
    isGlyphHidden={false}
    variant="error"
  />
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render with custom props 1`] = `
<div
  className="container testClassName"
>
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    />
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render with minimal props 1`] = `
<div
  className="container"
>
  <Scroll>
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    />
  </Scroll>
</div>
`;

exports[`ApplicationPageStatusProvider Snapshots should render with optional props 1`] = `
<div
  className="container"
>
  <Scroll
    refCallback={[MockFunction]}
  >
    <ContextProvider
      value={
        Object {
          "hide": [Function],
          "show": [Function],
        }
      }
    >
      <div>
        Test child
      </div>
    </ContextProvider>
  </Scroll>
</div>
`;
